python/ray/serve/tests/unit/test_metrics_utils.py

```diff
@@ -0,0 +1,230 @@
+            metrics_pusher.start()
+
+    def test_basic(self):
+        start = 0
+        timer = MockTimer(start)
+
+        with patch("time.time", new=timer.time), patch(
+            "time.sleep", new=timer.realistic_sleep
+        ):
+            counter = {"val": 0}
+            result = {}
+            expected_result = 20
+
+            def task(c, res):
+                timer.realistic_sleep(0.001)
+                c["val"] += 1
+                # At 10 seconds, this task should have been called 20 times
+                if timer.time() >= 10 and "val" not in res:
+                    res["val"] = c["val"]
+
+            metrics_pusher = MetricsPusher()
+            metrics_pusher.register_or_update_task(
+                "basic_task", lambda: task(counter, result), 0.5
+            )
+
+            metrics_pusher.start()
+            # This busy wait loop should run for at most a few hundred milliseconds
+            # The test should finish by then, and if the test fails this prevents
+            # an infinite loop
+            for _ in range(10000000):
+                if "val" in result:
+                    assert result["val"] == expected_result
+                    break
+
+            assert result["val"] == expected_result
+
+    def test_multiple_tasks(self):
+        start = 0
+        timer = MockTimer(start)
+
+        with patch("time.time", new=timer.time), patch(
+            "time.sleep", new=timer.realistic_sleep
+        ):
+            counter = {"A": 0, "B": 0, "C": 0}
+            result = {}
+            expected_results = {"A": 35, "B": 14, "C": 10}
+
+            def task(key, c, res):
+                time.sleep(0.001)
+                c[key] += 1
+                # Check for how many times this task has been called
+                # At 7 seconds, tasks A, B, C should have executed 35, 14, and 10
+                # times respectively.
+                if timer.time() >= 7 and key not in res:
+                    res[key] = c[key]
+
+            metrics_pusher = MetricsPusher()
+            # Each task interval is different, and they don't divide each other.
+            metrics_pusher.register_or_update_task(
+                "increment_A", lambda: task("A", counter, result), 0.2
+            )
+            metrics_pusher.register_or_update_task(
+                "increment_B", lambda: task("B", counter, result), 0.5
+            )
+            metrics_pusher.register_or_update_task(
+                "increment_C", lambda: task("C", counter, result), 0.7
+            )
+            metrics_pusher.start()
+
+            # Check there are three results set and all are expected.
+            def check_results():
+                for key in result.keys():
+                    assert result[key] == expected_results[key]
+                assert len(result) == 3
+                return True
+
+            wait_for_condition(check_results, timeout=20)
+
+    def test_update_task(self):
```

Comment:
This is the only new test, the rest of the tests in this file were moved with no changes.